src/zarr/api/asynchronous.py

```diff
@@ -0,0 +1,925 @@
+from __future__ import annotations
+
+import asyncio
+import warnings
+from collections.abc import Iterable
+from typing import Any, Literal, Union, cast
+
+import numpy as np
+import numpy.typing as npt
+
+from zarr.abc.codec import Codec
+from zarr.array import Array, AsyncArray
+from zarr.buffer import NDArrayLike
+from zarr.chunk_key_encodings import ChunkKeyEncoding
+from zarr.common import JSON, MEMORY_ORDER, ChunkCoords, OpenMode, ZarrFormat
+from zarr.group import AsyncGroup
+from zarr.metadata import ArrayV2Metadata, ArrayV3Metadata
+from zarr.store import (
+    StoreLike,
+    make_store_path,
+)
+
+ShapeLike = tuple[int, ...]  # TODO: support int for shape
+# TODO: this type could use some more thought, noqa to avoid "Variable "asynchronous.ArrayLike" is not valid as a type"
+ArrayLike = Union[AsyncArray | Array | npt.NDArray[Any]]  # noqa
+PathLike = str
+
+
+def _get_shape_chunks(a: ArrayLike | Any) -> tuple[ShapeLike | None, ChunkCoords | None]:
+    """helper function to get the shape and chunks from an array-like object"""
+    shape = None
+    chunks = None
+
+    if hasattr(a, "shape") and isinstance(a.shape, tuple):
+        shape = a.shape
+
+        if hasattr(a, "chunks") and isinstance(a.chunks, tuple) and (len(a.chunks) == len(a.shape)):
+            chunks = a.chunks
+
+        elif hasattr(a, "chunklen"):
+            # bcolz carray
+            chunks = (a.chunklen,) + a.shape[1:]
+
+    return shape, chunks
+
+
+def _like_args(a: ArrayLike, kwargs: dict[str, Any]) -> None:
```

Comment:
instead of mutating kwargs, can we copy it, and make changes to the copy, and return it?

Reply:
Good call. This function was coped from v2 but is now refactored along the lines of what you were suggesting.